registration/app/registration/services/fcm/FcmRegistrar.scala

```diff
@@ -0,0 +1,160 @@
+    def responseToInstance(response: WSResponse): Either[ProviderError, Instance] = response.status match {
+      case 200 => response.json.validate[Instance].fold(errors => Left(FcmProviderError(errors.mkString)), Right.apply)
+      case 404 => Left(InstanceNotFound)
+      case status => Left(FcmProviderError(s"Unable to fetch topics, got status $status back from the server"))
+    }
+
+    val response = ws.url(s"$instanceIdService/iid/info/${deviceToken.fcmToken}")
+      .addQueryStringParameters("details" -> "true")
+      .addHttpHeaders("Authorization" -> s"key=${configuration.firebaseServerKey}")
+      .get()
+
+    response.map(responseToInstance)
+  }
+
+  private def executeFirebaseTopicOperation[A](description: String)(f: FirebaseMessaging => TopicManagementResponse): RegistrarResponse[Unit] = {
+    val response = Future(f(firebaseMessaging))(fcmExecutionContext)
+    response.map { r =>
+      if (r.getSuccessCount != 1) {
+        val errors = r.getErrors.asScala.toList.map(_.getReason).mkString(",")
+        logger.error(s"Failed to $description, encountered following errors: $errors")
+        Left(FcmProviderError(s"Failed to $description, encountered following errors: $errors"))
+      } else Right(())
+    } recover {
+      case NonFatal(e) =>
+        logger.error(s"Failed to $description", e)
+        Left(FcmProviderError(s"Failed to $description: ${e.getMessage}"))
+    }
+  }
+
+  private def subscribeToTopic(deviceToken: DeviceToken)(topic: Topic): RegistrarResponse[Unit] = {
+    executeFirebaseTopicOperation(s"subscribe to topic ${topic.toFirebaseString}") {
+      _.subscribeToTopic(List(deviceToken.fcmToken).asJava, topic.toFirebaseString)
+    }
+  }
+
+  private def unsubscribeFromTopic(deviceToken: DeviceToken)(topic: Topic): RegistrarResponse[Unit] = {
+    executeFirebaseTopicOperation(s"unsubscribe from topic ${topic.toFirebaseString}") {
+      _.unsubscribeFromTopic(List(deviceToken.fcmToken).asJava, topic.toFirebaseString)
+    }
+  }
+
+  override def register(deviceToken: DeviceToken, registration: Registration): RegistrarResponse[RegistrationResponse] = {
+    def handleOperationError(results: Set[Either[ProviderError, Unit]]): Either[ProviderError, Unit] = {
+      val errors = results.collect { case Left(error) => error }
+      if (errors.isEmpty) {
+        Right(())
+      } else {
+        Left(FcmProviderError(s"Unable to update the registration, encountered errors: ${errors.mkString(", ")}"))
+      }
+    }
+
+    def forEachTopic(topics: Set[Topic])(f: Topic => RegistrarResponse[Unit]): RegistrarResponse[Unit] = {
+      Future.sequence(topics.map(f)).map(handleOperationError)
+    }
+
+    def listExistingTopics: RegistrarResponse[Set[Topic]] = {
+      fetchInstance(deviceToken).map {
+        case Right(instance) => Right(instance.topics.toSet)
+        case Left(InstanceNotFound) => Right(Set())
+        case Left(error) => Left(error)
+      }
+    }
+
+    val result = for {
```

Comment:
Is this safe if the server restarts?

Reply:
Connections are drained before the server is shut down so on that aspect it should be fine.

Even if the JVM explodes mid air, or if something else happen, the user might end up in an unstable position, which will be fix next time they try to register again (triggered every two weeks client side).

But assuming the server and the load balancer do behave correctly, it should return a 500, which will prompt the client to retry immediately up to three times.